### ISSUE TITLE
Update product label on propal/invoice/aso

### ISSUE THREAD

**Issue by rycks** [Dolibarr/dolibarr#23611]
### Bug

Two time this week some customers says "why we could not update product label on propal / invoice like description ?"
Same problem as described on french forum (https://www.dolibarr.fr/forum/t/modification-article/24307)

- first situation is : product on database is "generic product" and user change label on propal. Why do not use free lines on propal ? to get accounting numbers and other configuration.
- second one is : i do not have enought time to update my products lists, then i just want to make a generic product like "keyboard" and update the label on propal (like explained on french forum)

**Comment by rycks** [comment]
What we wish:

![image](https://user-images.githubusercontent.com/1468823/213121772-f8c1b54a-5f46-452d-9a76-15c5d15c0fbc.png)

**Comment by eldy** [comment]
The idea was that: if we update a label of a product (typo error for example), it is propagated everywhere with no need to update manually all its open documents. 
Why ?

Need A: In 90% of known use cases when a user do  a changes on the label of a product, he want to have the information propagated on non yet validated documents (because most of such changes are typo on label of product, a fix in translation, a fix of the accountancy code, or the fix of the image, ...). Note that if we change the price on product, this does not have effect on already validated documents: Propagation is never done for prices, new prices will appears only when doing a new documents.

Need B: In the 10% of cases where we want to have the change on product side not propagated, this is often because the ref and the product description is completely different. In such a case, it means it is a different product, so it is a bad practice to update an old product for that (because the accountancy code may also change, the custom code, the description, and statistics we want later will also be wrong). So a good practice is to create another product and close the old one (closing = setting status to "Not on sell", so it will disappear from the default views of products). Note: Closing a product to open a new one can be done easily by "Cloning" the old one if we want to reuse most information of old one.
So, both need A and B are currently covered by Dolibarr.

There is however a non common use case that is not covered: Some users want to change the label or the description of a product (the product is different and old one does not exist anymore) but don't want to create another product ref for that, and don't want to close the old one. This use case is currently not covered by Dolibarr. The reason for this use case is not yet clear, may be users don't want to increase the number of product ref thinking this can have performance troubles due to a high number of product references and prefer to no follow historic changes (in reality there is no trouble to manage millions of products ref and keeping an history of each product change, because only "on sale" products are visible by default, others are just "archive"). 

For the need to have "generic" product, the feature to enter "free" product was introduced for that. What is the goal of using a predefined product ref if label of product is modified and different ?  If a label changes for a different reason of a typo error or complement, it means surely that the product is different and stats on product ref need to be different too. So why not just using the "free product" line if the need is to be free in the label/description ? I imagine that the goal to keep using predefined product is to benefit the "predefined description". In such a case, it means product is a generic "example" product. So why not entering a generic label on the example product, use it as a predefined product and modify the label and the detail line to set its specific value just after adding the pre-selected product ?

Of course, this explains how it works today, not how it will work in future, this will depends on contributions received...

Note 1: When using the module BlockedLog (unalterable logs), a duplication of the reference and the label of products is stored into the table of unalterable log (table llx_blockedlog, this is a different table). Once recorded, this will never and can never be modified.
Note 2: When using a online signature of proposal, when a customer has signed a proposal, the document is archived with a date, so in this case whatever is the change done in your database (typo or not), the signed PDF will always remain unchanged too. You can, if you want, regenerate the document and ask your customer to resign it if you want  with the fixes, but the old signed document will remain untouched.

**Comment by rycks** [comment]
waw, product update -> propagate label update everywhere even on validated propal, i just discovered that now :-(

i made a propal, send it to my customer, then somebody update product label on products page and my propal is different than the customer has ... hmmm that's a bad thing no ?

for the other part of your message i agree but people are like they are :) and most of the time accounting code is not different because we make generic products like
- laptop
- ssd
- printer
- mouse
- keyboard

and so on, then when seller have to make a propal he just take "ssd" in the list and put "SSD : Samsung xxx" in label and maybe more data in description. Most of the time sellers don't care about accounting problems and they could use free lines but with free lines accounter lost his time to make the job "what is samsung128 ? hmmm that's a ssd, ok accounting number is ... ho let me check, that was a customer from outside europe, that's not the good number" etc.

**Comment by eldy** [comment]
I replaced the flag "Bug" into "Feature request" because this was a discussion on how to support product snapshot on documents (not yet implemented as explained), to avoid to have the thread closed automatically by the bot...

**Comment by hregis** [comment]
@rycks @eldy le champ "label" sur les lignes je l'avais rajouté il y a longtemps, ça permettait justement de pouvoir personnaliser le libellé du produit si besoin... Laurent souviens toi tu n'étais pas d'accord avec ce principe et tu as supprimé cette option... est-ce qu'on pourrait se mettre d'accord une bonne fois pour toute et éviter d'avancer et de reculer en permanence ?! ;-) :-)

**Comment by eldy** [comment]
> @rycks @eldy le champ "label" sur les lignes je l'avais rajouté il y a longtemps, ça permettait justement de pouvoir personnaliser le libellé du produit si besoin... Laurent souviens toi tu n'étais pas d'accord avec ce principe et tu as supprimé cette option... est-ce qu'on pourrait se mettre d'accord une bonne fois pour toute et éviter d'avancer et de reculer en permanence ?! ;-) :-)

True, the feature you mention was hidden for several reasons: 
* Implementation was not complete (work for some object for example and not other, sometimes in edit mode sometimes in create or the opposite. It was before the edit/create code was mutualized in tpl files).
* Problem when using multilang (feature broken)
* Users complains about why they have to change tons of documents when they fix a typo and need to remove and reinsert a line in a document when they have fixed something in the product to have change effective. Having such feedbacks on something that was working previously is enough to say that implementation must probably be thought differently, before being officially introduced. It seems a very large number of users was expecting the current way of working, we must not ignore that before changing something.
All this points are still true.

Because the troubles and complains of users has stopped after the hide of the feature, this topic was forgotten until a new PR appears recently on this.

My understanding here is that the expected feature is not to be able to customize a label but to be able to have product informations archived at document level so when we change one data on a product, user can choose to propagate or not the change on document (not sure this is specific to product labels, this may be true for description, the custom code (code douane), translations, and may be other fields.

**Comment by hregis** [comment]
@eldy @rycks I don't think we should give the possibility of modifying a "document", quote, invoice, etc... as soon as this document is validated. An object must remain in the state in which it was created. You do not agree ?

My point of view is not important. The point of view of users should be. As I said, a lot of users complains when propagation of a typo fix was not working or must be done manually on each document (and this is a frequent case why product labels are modified). So doing differently must be done taking this is account...

**Comment by hregis** [comment]
@eldy @rycks it would be good to record the label and description of a product in the line of an object so that this line is not modified if the original product is modified

**Comment by hregis** [comment]
@eldy @rycks après on peut aller plus loin... toutes informations d'un produit (label, desc, barcode, etc...) devraient être figés sur le document ! c'est le soucis actuellement... tu modifies quoi que ce soit sur un produit, ou quoi que ce soit d'autres... on se base sur l'id... si les données de l'id sont modifiés, tout ce qui se rapporte à cet id est modifié...

**Comment by hregis** [comment]
@eldy @rycks on avait franchi ce cap avec le fork et une base nosql... chaque document contenait les coordonnées du moment du vendeur et de l'acheteur, et chaque ligne contenait les données du moment du produit...
faire ça avec une base de données relationnelle c'est une usine à gaz... :-)

**Comment by hregis** [comment]
@eldy @rycks je vais prêcher pour ma paroisse, le nosql est un concept complètement différent... c'est le code qui crée la structure de la base, à l'inverse des bases relationnelles... chaque tiers, devis, facture, etc... est un "document"... un document c'est du json... tu peux mettre ce que tu veux dedans.... c'est le code qui le détermine, et le code c'est toi qui l'écrit ! ;-) 
pas de requêtes sql... tu charges le document, tu le modifies, et tu l’enregistres... 
on avait enlevé 50% du code de Dolibarr !
Amen ! ;-)
